lib/std/channel.nim

```diff
@@ -0,0 +1,702 @@
+  result = true
+
+proc channel_close_mpsc(chan: ChannelRaw): bool =
+  # Unsynchronized
+  assert not chan.isNil
+
+  if chan.isClosed():
+    # Already closed
+    result = false
+  else:
+    chan.closed.store(true, moRelaxed)
+    result = true
+
+proc channel_open_mpsc(chan: ChannelRaw): bool =
+  # Unsynchronized
+  assert not chan.isNil
+
+  if not chan.isClosed():
+    # Already open
+    result = false
+  else:
+    chan.closed.store(false, moRelaxed)
+    result = true
+
+# SPSC Channels (Single-Producer Single-Consumer)
+# ----------------------------------------------------------------------------------
+
+proc channel_send_unbuffered_spsc(chan: ChannelRaw, data: sink pointer, size: int32, nonBlocking: bool): bool =
+  if nonBlocking and chan.isFullUnbuf:
+    return false
+
+  while chan.isFullUnbuf:
+    cpuRelax()
+
+  assert chan.isEmptyUnbuf
+  assert size <= chan.itemsize
+  copyMem(chan.buffer, data, size)
+
+  fence(moSequentiallyConsistent)
+
+  chan.head = 1
+  signal(chan.notEmptyCond)
+  result = true
+
+proc channel_send_spsc(chan: ChannelRaw, data: sink pointer, size: int32, nonBlocking: bool): bool =
+  assert not chan.isNil
+  assert not data.isNil
+
+  if chan.isUnbuffered():
+    return channel_send_unbuffered_spsc(chan, data, size, nonBlocking)
+
+  if nonBlocking and chan.isFull():
+    return false
+
+  while chan.isFull():
+    cpuRelax()
+
+  assert not chan.isFull()
+  assert size <= chan.itemsize
+  copyMem(chan.buffer[chan.tail * chan.itemsize].addr, data, size)
+
+  let newTail = chan.tail.incmod(chan.size)
+
+  fence(moSequentiallyConsistent)
+
+  chan.tail = newTail
+  signal(chan.notEmptyCond)
+  result = true
+
+proc channel_recv_spsc(chan: ChannelRaw, data: pointer, size: int32, nonBlocking: bool): bool =
+  # Cannot be inline due to function table
+  channel_recv_mpsc(chan, data, size, nonBlocking)
+
+proc channel_close_spsc(chan: ChannelRaw): bool =
+  # Unsynchronized
+  assert not chan.isNil
+
+  if chan.isClosed():
+    # Already closed
+    result = false
+  else:
+    chan.closed.store(true, moRelaxed)
+    result = true
+
+proc channel_open_spsc(chan: ChannelRaw): bool =
+  # Unsynchronized
+  assert not chan.isNil
+
+  if not chan.isClosed():
+    # Already open
+    result = false
+  else:
+    chan.closed.store(false, moRelaxed)
+    result = true
+
+# "Generic" dispatch
+# ----------------------------------------------------------------------------------
+
+const
+  send_fn = [
+    Mpmc: sendMpmc,
+    Mpsc: channel_send_mpsc,
+    Spsc: channel_send_spsc
+  ]
+
+  recv_fn = [
+    Mpmc: recvMpmc,
+    Mpsc: channel_recv_mpsc,
+    Spsc: channel_recv_spsc
+  ]
+
+  close_fn = [
+    Mpmc: channel_close_mpmc,
+    Mpsc: channel_close_mpsc,
+    Spsc: channel_close_spsc
+  ]
+
+  open_fn = [
+    Mpmc: channel_open_mpmc,
+    Mpsc: channel_open_mpsc,
+    Spsc: channel_open_spsc
+  ]
+
+proc channel_send(chan: ChannelRaw, data: sink pointer, size: int32, nonBlocking: bool): bool {.inline.} =
```

Comment:
style: `channelSend` (ditto for style with other symbols)

or is that intended?

Reply:
I'm not sure, but snake_case also appear in `std/atomics`

Reply:
Don't use `snake_case`.

Reply:
only `tcc_cas` and 6 ATOMIC_XXX symbols, all the rest is camlCase (including vast majority of stdib)

Reply:
fixed